pkg/cache/cluster.go

```diff
@@ -1004,6 +1007,94 @@ func (c *clusterCache) IterateHierarchy(key kube.ResourceKey, action func(resour
 	}
 }
 
+// IterateHierarchy iterates resource tree starting from the specified top level resources and executes callback for each resource in the tree
+func (c *clusterCache) IterateHierarchyV2(keys []kube.ResourceKey, action func(resource *Resource, namespaceResources map[kube.ResourceKey]*Resource) bool) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	keysPerNamespace := make(map[string][]kube.ResourceKey)
+	for _, key := range keys {
+		keysPerNamespace[key.Namespace] = append(keysPerNamespace[key.Namespace], key)
```

Comment:
The existing `IterateHierarchy` method short-circuits the whole function if the key doesn't exist in `c.resources`. We should keep that optimization here.

Reply:
I have a similar block below
```
	for _, key := range namespaceKeys {
			res, ok := c.resources[key]
			if !ok {
				continue
			}
```
But yeah, I guess I can skip it here.

Reply:
Updated the PR, hopefully addressing this.

Reply:
Yep, that check was after building the graph, I think we should do the check before.

Reply:
Update looks good to me!